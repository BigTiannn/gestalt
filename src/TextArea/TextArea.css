.textArea {
  composes: lightGrayBg darkGray from "../Colors.css";
  composes: bold text-md from global;
  composes: border-box from global;
  composes: col-12 from global;
  composes: px1 py2 from global;
  composes: rounded from global;
  cursor: text;
  outline: 0;
  resize: none;
}

.normal {
  composes: border from global;
}

.textArea:focus {
  background-color: #fff;
}

.errored {
  border: 1px solid #e3780c;
}

.errored:focus {
  background-color: #fff;
}
