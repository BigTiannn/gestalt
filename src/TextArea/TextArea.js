// @flow

import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import ErrorFlyout from '../ErrorFlyout/ErrorFlyout';
import styles from './TextArea.css';

type State = {
  focused: boolean,
  errorIsOpen: boolean,
};

export default class TextArea extends Component {
  static propTypes = {
    errorMessage: PropTypes.string,
    id: PropTypes.string.isRequired,
    onBlur: PropTypes.func,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    placeholder: PropTypes.string,
    value: PropTypes.string,
  };

  state:State = {
    focused: false,
    errorIsOpen: false,
  };

  props: {|
    errorMessage?: string,
    id: string,
    name?: string,
    onBlur?: (value: string) => void,
    onChange: (value: string) => void,
    onFocus?: (value: string) => void,
    placeholder?: string,
    value?: string,
  |};

  handleChange = (e: Event) => {
    if (e.target instanceof HTMLTextAreaElement) {
      this.props.onChange(e.target.value);
    }
  }

  handleBlur = (e: Event) => {
    if (this.props.errorMessage) {
      this.setState({ errorIsOpen: false });
    }
    if (e.target instanceof HTMLTextAreaElement && this.props.onBlur) {
      this.props.onBlur(e.target.value);
    }
  }

  handleFocus = (e: Event) => {
    if (this.props.errorMessage) {
      this.setState({ errorIsOpen: true });
    }
    if (e.target instanceof HTMLTextAreaElement && this.props.onFocus) {
      this.props.onFocus(e.target.value);
    }
  }

  render() {
    const {
      errorMessage,
      id,
      name,
      placeholder,
      value,
    } = this.props;

    const classes = classnames(styles.textArea, {
      [styles.normal]: !errorMessage,
      [styles.errored]: errorMessage,
    });

    const textArea = () => (
      <textarea
        aria-describedby={errorMessage && this.state.focused ? `${id}-gestalt-error` : null}
        aria-invalid={errorMessage ? 'true' : 'false'}
        className={classes}
        id={id}
        name={name}
        onBlur={this.handleBlur}
        onChange={this.handleChange}
        onFocus={this.handleFocus}
        placeholder={placeholder}
        rows={3}
        value={value}
      />
    );

    return errorMessage ? (
      <ErrorFlyout
        id={`${id}-gestalt-error`}
        idealDirection="right"
        isOpen={this.state.errorIsOpen}
        message={errorMessage}
        onDismiss={() => this.setState({ errorIsOpen: false })}
        size="sm"
        trigger={textArea()}
      />
    ) : textArea();
  }
}
