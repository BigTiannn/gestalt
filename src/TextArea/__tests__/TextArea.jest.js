/* eslint-env jest */
/* eslint import/imports-first: 0 */
jest.unmock('../TextArea');

import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import TextArea from '../TextArea';

describe('TextField', () => {
  it('Renders an ErrorFlyout if one is passed in', () => {
    const wrapper = shallow(
      <TextArea
        errorMessage="test"
        id="test"
        onChange={jest.fn()}
      />
    );
    expect(wrapper.find('ErrorFlyout').length).toEqual(1);
  });

  it('Does not render an ErrorFlyout when errorMessage is null', () => {
    const wrapper = shallow(
      <TextArea
        id="test"
        onChange={jest.fn()}
      />
    );
    expect(wrapper.find('ErrorFlyout').length).toEqual(0);
  });

  it('TextArea normal', () => {
    const tree = create(
      <TextArea
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('TextArea with error', () => {
    const tree = shallow(
      <TextArea
        errorMessage="error message"
        id="test"
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    ).html();
    expect(tree).toMatchSnapshot();
  });
});
