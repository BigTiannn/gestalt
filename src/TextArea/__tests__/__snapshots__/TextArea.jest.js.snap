exports[`TextField TextArea normal 1`] = `
<textarea
  aria-describedby={null}
  aria-invalid="false"
  className="textArea normal"
  id="test"
  name={undefined}
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder={undefined}
  rows={3}
  value={undefined} />
`;

exports[`TextField TextArea with error 1`] = `"<div><div><textarea aria-invalid=\"true\" class=\"textArea errored\" id=\"test\" rows=\"3\"></textarea></div><div></div></div>"`;
